orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java

```diff
@@ -194,7 +196,50 @@ private RollbackDetails findBestCandidate(
         previousImageRollbackSupport.getImageDetailsFromEntityTags(
             stageData.cloudProvider, stageData.credentials, region, serverGroupToRollBack.name);
 
-    return getBestCandidate(cluster, region, serverGroupToRollBack, candidates, imageDetails);
+    return getOriginalServerGroup(stage)
+        .map(
+            serverGroupToRestore ->
+                getRollbackDetails(serverGroupToRollBack.name, serverGroupToRestore, imageDetails))
+        .orElseGet(
+            () ->
+                getBestCandidate(cluster, region, serverGroupToRollBack, candidates, imageDetails));
+  }
+
+  private RollbackDetails getRollbackDetails(
+      String serverGroupToRollBack, String serverGroupToRestore, ImageDetails imageDetails) {
+
+    Map<String, String> context =
+        ImmutableMap.<String, String>builder()
+            .put("rollbackServerGroupName", serverGroupToRollBack)
+            .put("restoreServerGroupName", serverGroupToRestore)
+            .build();
+
+    return new RollbackDetails(
+        EXPLICIT, context, imageDetails.getImageName(), imageDetails.getBuildNumber());
+  }
+
+  /** Return the name of the original server group we should roll back to */
+  private Optional<String> getOriginalServerGroup(StageExecution stage) {
+    return Optional.ofNullable(stage.findAncestor(isType("createServerGroup")))
+        .map(StageExecution::getContext)
+        .map(this::getSourceServerGroup);
+  }
+
+  /**
+   * Given context info from a createServerGroup stage, extract the field:
+   *
+   * <p>{"source": {"serverGroupName": "`ASG-NAME-HERE`" }}
+   */
+  @Nullable
+  private String getSourceServerGroup(Map<String, Object> contextFromCreateServerGroupStage) {
+    Object source = contextFromCreateServerGroupStage.get("source");
+    if (source instanceof Map) {
+      Object serverGroupName = ((Map<?, ?>) source).get("serverGroupName");
+      if (serverGroupName instanceof String) {
```

Comment:
Check for `null`? I guess a simpler way to do this, since you're already calling `toString()` on the value below, would be to just replace this with:
```suggestion
      if (serverGroupName != null) {
```

Reply:
The problem with just checking for null is that if this points to something that isn't a string (e.g., an object,list,number), then this call will succeed, but the resulting string almost certainly won't be a valid server group name. I'd prefer to return `null` in that case.

Now that I'm writing this, I think that I should add a log statement upon returning null here, since we always expect the "source" field to be present.

Reply:
Added a log statement in 878bee228